Cannot paste text from PDF in Safari to Keynote

aprotyas opened the pull request:
#### 9f8355e31ea09f1bc1a9c44d46e779ff754e0610
<pre>
Cannot paste text from PDF in Safari to Keynote
<a href="https://bugs.webkit.org/show_bug.cgi?id=282715">https://bugs.webkit.org/show_bug.cgi?id=282715</a>
<a href="https://rdar.apple.com/136075998">rdar://136075998</a>

Reviewed by Wenson Hsieh.

Both -[PDFSelection {webArchive, html}] have been deprecated, with the
former returning nil unconditionally and the latter being generally
broken in macOS 15. This meant that we were advertisting pasteboard data
types with empty items, which mislead clients and resulted in symptoms
captured by the bug title.

This patch does the following to address the bug:
1. Drops our call to -[PDFSelection webArchive], since we would never
   have any  meaningful web archive content to put on the pasteboard.
2. Adopts -[PDFSelection htmlData] in lieu of -[PDFSelection html].
3. Advertises RTF data on the pasteboard by adopting -[PDFSelection
   rtfData].

We also:
1. Reorder the order of the pasteboard items we are inserting in the
   plugin so that higher fidelity data (HTML/RTF) appears before plain
   string data.
2. Change the signature of PDFPluginBase::writeItemsToPasteboard() so
   that callers cannot mistakenly provide mismatching arrays for items
   and types.

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(-[WKPDFLayerControllerDelegate writeItemsToPasteboard:withTypes:]):
(WebKit::PDFPlugin::handleEditingCommand):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::writeItemsToPasteboard const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::performCopyLinkOperation const):
(WebKit::htmlDataFromSelection):
(WebKit::UnifiedPDFPlugin::performCopyEditingOperation const):
(WebKit::UnifiedPDFPlugin::takeFindStringFromSelection):

Canonical link: <a href="https://commits.webkit.org/286274@main">https://commits.webkit.org/286274@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b3f0449c925a2c49ca3504668b6cef0a5e1588d5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75418 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/80 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28268 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79894 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26682 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/80 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2649 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59186 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17397 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78485 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/80 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64798 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39544 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/80 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25010 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/80 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22626 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81376 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2757 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1735 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/67430 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2908 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64780 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66720 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16608 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10671 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/8826 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2714 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5535 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2739 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3674 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2746 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
```diff
@@ -50,6 +50,7 @@
 #include <wtf/TypeTraits.h>
 #include <wtf/WeakPtr.h>
 
+OBJC_CLASS NSData;
```
Do you need `OBJC_CLASS NSString;` as well?

aprotyas replied:
I do, I'll add it so we don't implicitly depend on a transitive `Foundation/Foundation.h` import.

whsieh commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2567,11 +2567,12 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+    RetainPtr urlData = [[url absoluteString] dataUsingEncoding:NSUTF8StringEncoding];
+    Vector<PasteboardItem> pasteboardItems {
+        { urlData, NSPasteboardTypeString },
+        { urlData, NSPasteboardTypeHTML },
```
Would this write HTML data for a copied URL? I think we'd probably either just want the plain string, or use `NSPasteboardTypeURL`

aprotyas replied:
Yeah this seemed odd to be, but I didn't question it since we migrated this from legacy PDF plugin.

I do think if we advertise a second data type here then `NSPasteboardTypeURL` is the better fit. I'll make sure that change doesn't break anything and then update this line.

aprotyas committed:
Cannot paste text from PDF in Safari to Keynote
https://bugs.webkit.org/show_bug.cgi?id=282715
rdar://136075998

Reviewed by Wenson Hsieh.

Both -[PDFSelection {webArchive, html}] have been deprecated, with the
former returning nil unconditionally and the latter being generally
broken in macOS 15. This meant that we were advertisting pasteboard data
types with empty items, which mislead clients and resulted in symptoms
captured by the bug title.

This patch does the following to address the bug:
1. Drops our call to -[PDFSelection webArchive], since we would never
   have any  meaningful web archive content to put on the pasteboard.
2. Adopts -[PDFSelection htmlData] in lieu of -[PDFSelection html].
3. Advertises RTF data on the pasteboard by adopting -[PDFSelection
   rtfData].

We also:
1. Reorder the order of the pasteboard items we are inserting in the
   plugin so that higher fidelity data (HTML/RTF) appears before plain
   string data.
2. Change the signature of PDFPluginBase::writeItemsToPasteboard() so
   that callers cannot mistakenly provide mismatching arrays for items
   and types.

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(-[WKPDFLayerControllerDelegate writeItemsToPasteboard:withTypes:]):
(WebKit::PDFPlugin::handleEditingCommand):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::writeItemsToPasteboard const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::performCopyLinkOperation const):
(WebKit::htmlDataFromSelection):
(WebKit::UnifiedPDFPlugin::performCopyEditingOperation const):
(WebKit::UnifiedPDFPlugin::takeFindStringFromSelection):

Canonical link: https://commits.webkit.org/286274@main
